File: Leetcode/Python/1038. Binary Search Tree to Greater Sum Tree/solution.py
# https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    total = 0

    def bstToGst(self, root: TreeNode) -> TreeNode:
        if not root:
            return 0

        # BST has bigger nodes on the right side than the current node
        self.bstToGst(root.right)
        self.total += root.val
        root.val = self.total
        # After add parent node and right side nodes to total than add left side nodes
        self.bstToGst(root.left)

        return root

File: Leetcode/Java/1038. Binary Search Tree to Greater Sum Tree/solution.java
// https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int total = 0;

    public TreeNode bstToGst(TreeNode root) { // In-order
        if (root == null) {
            return null;
        }

        bstToGst(root.right);
        total += root.val;
        root.val = total;
        bstToGst(root.left);

        return root;
    }
}